File: src/main/java/com/freya02/botcommands/api/parameters/ParameterResolverFactory.java
package com.freya02.botcommands.api.parameters;

import com.freya02.botcommands.api.core.utils.ReflectionUtils;
import kotlin.reflect.KClass;
import org.jetbrains.annotations.NotNull;

/**
 * Factory for {@link ParameterResolver parameter resolvers}.
 *
 * @see ParameterResolver
 */
public abstract class ParameterResolverFactory<T extends ParameterResolver<?, R>, R> {
	private final KClass<T> resolverType;
	private final KClass<R> jvmErasure;

	/**
	 * Constructs a new parameter resolver factory
	 *
	 * @param resolverType Class of the resolver
	 * @param clazz Class of the parameter being resolved
	 */
	public ParameterResolverFactory(Class<T> resolverType, @NotNull Class<R> clazz) {
		this.resolverType = ReflectionUtils.toKotlin(resolverType);
		this.jvmErasure = ReflectionUtils.toKotlin(clazz);
	}

	/**
	 * Constructs a new parameter resolver factory
	 *
	 * @param resolverType Class of the resolver
	 * @param clazz Class of the parameter being resolved
	 */
	public ParameterResolverFactory(KClass<T> resolverType, @NotNull KClass<R> clazz) {
		this.resolverType = resolverType;
		this.jvmErasure = clazz;
	}

	public KClass<T> getResolverType() {
		return resolverType;
	}

	public KClass<R> getJvmErasure() {
		return jvmErasure;
	}

	@NotNull
	public abstract T get(@NotNull ParameterWrapper parameter);

	@SuppressWarnings({"unchecked", "rawtypes"})
	public static <T extends ParameterResolver<?, R>, R> ParameterResolverFactory<T, R> singleton(T resolver) {
		return new ParameterResolverFactory<>(resolver.getClass(), ReflectionUtils.toJava(resolver.getJvmErasure())) {
			@NotNull
			@Override
			public T get(@NotNull ParameterWrapper parameter) {
				return resolver;
			}
		};
	}
}

File: BotCommands-core/src/main/kotlin/io/github/freya022/botcommands/api/parameters/ParameterResolverFactory.kt
package io.github.freya022.botcommands.api.parameters

import io.github.freya022.botcommands.api.core.reflect.ParameterWrapper
import io.github.freya022.botcommands.api.core.service.annotations.InterfacedService
import io.github.freya022.botcommands.api.core.service.annotations.Resolver
import io.github.freya022.botcommands.api.core.service.annotations.ResolverFactory
import io.github.freya022.botcommands.api.core.utils.allInterfaces
import io.github.freya022.botcommands.api.core.utils.isSubclassOf
import io.github.freya022.botcommands.api.core.utils.shortQualifiedName
import io.github.freya022.botcommands.api.core.utils.simpleNestedName
import io.github.freya022.botcommands.api.parameters.resolvers.IParameterResolver
import io.github.freya022.botcommands.internal.parameters.resolvers.ResolverMarker
import kotlin.reflect.KType

/**
 * Factory for [parameter resolvers][ParameterResolver].
 *
 * Your implementation needs to be annotated with [@ResolverFactory][ResolverFactory].
 *
 * ### How it works
 *
 * A factory determines if a given parameter is supported, if so, a parameter resolver will be created.
 *
 * If multiple factories return `true` in [isResolvable] for a given type,
 * the factory with the best [priority] is taken, if two with the same top priority exists, an exception is thrown.
 *
 * ### Use cases
 *
 * This is particularly useful if your parameter's type has generics, which you may read in [ParameterWrapper.type],
 * or if you need to read an annotation on the [parameter][ParameterWrapper.parameter]
 * to adjust the resolver's behavior, for example.
 *
 * In case you want to read generics, you can read them off the [KType] in [ParameterWrapper],
 * but you can only do that with Kotlin.
 *
 * In case you want to read the annotations, you can use the methods supplied by [ParameterWrapper].
 *
 * @see TypedParameterResolverFactory
 * @see ParameterResolver
 */
@InterfacedService(acceptMultiple = true)
abstract class ParameterResolverFactory {

    /**
     * List of types as strings that are supported by this resolver factory.
     *
     * This should be the types of what the returned parameter resolvers are capable of returning.
     *
     * This is only used for logging purposes.
     */
    abstract val supportedTypesStr: List<String>

    /**
     * List of resolvers supported by this resolver factory.
     *
     * These should only be types directly extending `BuiltinResolver` (an internal interface).
     */
    abstract val supportedResolvers: List<Class<out IParameterResolver<*>>>

    /**
     * The priority of this factory.
     *
     * When getting a resolver factory, the factory with the highest value that is [resolvable][isResolvable] is taken.
     *
     * If two factories with the same priority exist and are both resolvable, an exception is thrown.
     *
     * @see Resolver.priority
     */
    open val priority: Int get() = 0

    /**
     * Determines if a given parameter is supported.
     *
     * Out of all factories supporting the given [ResolverRequest.resolverType], and with the same [priority], only one must return `true`.
     *
     * This only runs for resolvers declared as supported by [supportedResolvers].
     */
    abstract fun isResolvable(request: ResolverRequest): Boolean

    /**
     * Returns a [ParameterResolver] for the given parameter.
     *
     * This is only called if [isResolvable] returned `true`.
     */
    abstract fun get(request: ResolverRequest): IParameterResolver<*>

    override fun toString(): String {
        return "ParameterResolverFactory(supportedResolvers=${supportedResolvers.map { it.simpleNestedName }}, supportedTypes=${supportedTypesStr})"
    }

    open fun toLogString(): String {
        return "${this.javaClass.shortQualifiedName} ; priority $priority (${supportedTypesStr.joinToString()})"
    }

    companion object {
        // TODO docs
        @JvmStatic
        fun inferSupportedResolversFrom(resolverType: Class<out IParameterResolver<*>>): List<Class<out IParameterResolver<*>>> {
            @Suppress("UNCHECKED_CAST")
            return resolverType
                .allInterfaces
                .filter { it.isSubclassOf<IParameterResolver<*>>() && ResolverMarker::class.java in it.interfaces }
                    as List<Class<out IParameterResolver<*>>>
        }

        @JvmSynthetic
        inline fun <reified T : IParameterResolver<*>> inferSupportedResolversFrom(): List<Class<out IParameterResolver<*>>> {
            return inferSupportedResolversFrom(T::class.java)
        }
    }
}
